test/op_shape_test.cpp

```diff
@@ -2181,6 +2183,33 @@ TEST_CASE(reshape_shape)
     }
 }
 
+TEST_CASE(reshape_nonstandard_unsqueeze)
+{
+    migraphx::shape input{migraphx::shape::float_type, {4, 24, 1, 1, 1}, {1, 4, 1, 1, 1}};
+    std::vector<std::size_t> lens = {4, 1, 3, 4, 2};
+    std::vector<int64_t> perm     = {4, 0, 1, 2, 3};
+    migraphx::shape output        = migraphx::shape::from_permutation(
+        migraphx::shape::float_type, lens, migraphx::invert_permutation(perm));
```

Comment:
> Is this from_permutation() with the invert_permutation equivalent to:

Its equivalent of a transpose with the same permutation preceding the reshape. 

> reorder_shape(shape{t, lens={1, 3, 4, 2, 4}}, perm={4, 0, 1, 2, 3})

This doesn't make clear the order of lens which is `4, 1, 3, 4, 2`.

> shape{t, lens={4, 1, 3, 4, 2}, strides = {1, 96, 32, 8, 4}}

This doesnt make clear the transpose that was applied to the shape.

Reply:
> This doesnt make clear the transpose that was applied to the shape. 

Input is already transposed. how would one calculate `perm` for the expected output shape ({4, 0, 1, 2, 3} in this case), if one were to write such test  ?

Reply:
The permutation of the input is `1, 0, 2, 3, 4`, and then as we collapse or expand dimensions from `4, 24, 1, 1, 1` to `4, 1, 3, 4, 2` we remove the collapsed dimensions or duplicate the expanded dimensions in the permutation which leads to `1, 0, 0, 0, 0` and then renumber the permutation to avoid dups or gaps which is `4, 0, 1, 2, 3`.

Reply:
okay. IMO following is a better check than trying to figure out what the perm would be. 
> shape{t, lens={4, 1, 3, 4, 2}, strides = {1, 96, 32, 8, 4}}?

> This doesn't make clear the transpose that was applied to the shape. 

Reshape is not doing any transpose. Input is already transposed.  Reshape is just taking one or more axes of the transposed shape and reinterpreting it as a different shape. 
Users don't need to be made aware if input itself is transposed. Users probably know it already since they provide input).

Reply:
> How do you figure out the strides without figuring out the permutation first?

 By knowing the access pattern. e.g. input has strides {1, 4, 1, 1, 1} with shape {4, 24, 1, 1, 1}.  Axes 2, 3 and 4 are fixed with 0 index. Therefore fast-moving axis is axis 1 with stride 4 resulting in strided buffer access and axes 0 is continuous access. 

Reshape maintains memory access patterns. 

If I were to reshape it to {4, 1, 3, 4, 2} then, I need to keep fast-moving axes at stride 4 that result in strides 
{x, y ,  4 * 2 * 4, 2* 4, 4}.  That completes access of 24 elements and then, axes 1 with shape 1 is fixed.  So doens't really matter what strides it has. axes 0 originally was contiguous memory access. The shape of it remains unchanged at 4. So final strides = {1, x, 32, 8, 4}.

Reply:
> Therefore fast-moving axis is axis 1 with stride 4

The fastest moving axis is 0. What is the fast-moving axis mean?

> If I were to reshape it to {4, 1, 3, 4, 2} then, I need to keep fast-moving axes at stride 4 that result in strides
{x, y , 4 * 2 * 4, 2* 4, 4}. That completes access of 24 elements and then, axes 1 with shape 1 is fixed. So doens't really matter what strides it has. axes 0 originally was contiguous memory access. The shape of it remains unchanged at 4. So final strides = {1, x, 32, 8, 4}.

Using the permutation seems way simpler. There is no need to calculate strides.

Reply:
> The fastest moving axis is 0. What is the fast-moving axis mean?

By fastest, i meant axis that would have index changes most frequently. 
> Using the permutation seems way simpler. There is no need to calculate strides. 

We should document that in that case, if I were to read this test case after 1 year, I would be puzzled where perm came from.